### ISSUE TITLE
Chef-14:  chef-zero / local-mode taking forever to depsolve on large cookbook sets

### ISSUE THREAD

**Issue by lamont-granquist** [chef/chef#7120]
I've got a good repro case, very unclear right now what is going on.

**Comment by lamont-granquist** [comment]
rolling back chef-zero to 13.1.0 exhibits the same problem, so it shouldn't be a chef-zero 14.0 code change.

using appbundle-updater, though, to install chef 13.8.5 on top of ruby-1.8.5 fixes it and my repro case goes from 60 seconds back down to 1 second, so it also appears to not be a ruby 2.5.1 problem.

**Comment by lamont-granquist** [comment]
chef 14.0.56 at b89bbe9fced92791da5cd647490bb31d55ac49df with ohai v14.0.2 is slow.

**Comment by lamont-granquist** [comment]
```ruby
            cookbook_entry = entry.children.select do |child|
              child.chef_object.full_name == "#{path[1]}-#{path[2]}" ||
                (child.chef_object.name.to_s == path[1] && child.chef_object.identifier == path[2])
            end[0]
```

That is the guilty line of code.

**Comment by lamont-granquist** [comment]
(i'm still have no idea why that particular line of code is so slow.

**Comment by lamont-granquist** [comment]
seems to be that `child.chef_object` takes 0.02sec times three calls, times as many cookbooks as there are in the solution, which for >10-ish cookbook adds up to >1 second, times it looks like as many cookbooks as there are in the solution again(?) for potentially minutes.  so it looks like it may be O(N^2) with a large constant.

**Comment by lamont-granquist** [comment]
kind of looks like `child.child_object` calls `Chef::ChefFS::ChefFSDataStore#get` which uses the CookbookVersionLoader and ultimately calls `Chef::Cookbook::CookbookVersionLoader#load_all_files` -- so in order to retrieve one entry we're looping over all the cookbooks again -- which I think makes this O(n^3)

**Comment by lamont-granquist** [comment]
actually not, its just O(n^2), but the outer loop is in the chef-client where it hits the successive CBV endpoints.  chef-zero can't easily cache across those.  each one of those now loops through every CBV looking for the one with the right name.

**Comment by lamont-granquist** [comment]
closed by #7143
